system_metrics_collector/include/moving_average_statistics/moving_average.hpp

```diff
@@ -0,0 +1,189 @@
+// Copyright 2019 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef MOVING_AVERAGE_STATISTICS__MOVING_AVERAGE_HPP_
+#define MOVING_AVERAGE_STATISTICS__MOVING_AVERAGE_HPP_
+
+
+#include <cmath>
+
+#include <algorithm>
+#include <mutex>
+#include <type_traits>
+
+#include "moving_average_statistics/types.hpp"
+
+
+/**
+ *  A class for calculating moving average statistics. This operates in constant memory and constant time. Note:
+ *  reset() must be called manually in order to start a new measurement window.
+ *
+ *  The statistics calculated are average, maximum, minimum, and standard deviation (population).
+ *  All are calculated online without storing the observation data. Specifically, the average is a running sum
+ *  and the variance is obtained by Welford's online algorithm
+ *  (reference: https://en.wikipedia.org/wiki/Algorithms_for_calculating_variance#Welford%27s_online_algorithm)
+ *  for standard deviation.
+ *
+ *  When statistics are not available, e.g. no observations have been made, NaNs are returned.
+ *
+ *  @tparam T the type to be observed. Note: this must conform to std::is_arithmetic.
+**/
+
+template<
+  class T,
+  typename = typename std::enable_if<std::is_arithmetic<T>::value, T>::type>
+class MovingAverageStatistics
+{
+public:
+  /**
+   *  MovingAverageStatistics object constructor.
+  **/
+  MovingAverageStatistics() = default;
+  ~MovingAverageStatistics() = default;
+
+  /**
+   *  Returns the arithmetic mean of all data recorded. If no observations have been made, returns NaN.
+   *
+   *  @return The arithmetic mean of all data recorded, or NaN if the sample count is 0.
+  **/
+  double average()
+  {
+    return getStatistics().average;
+  }
+  /**
+   *  Returns the maximum value recorded. If size of list is zero, returns NaN.
+   *
+   *  @return The maximum value recorded, or NaN if size of data is zero.
+  **/
+  double max()
+  {
+    return getStatistics().max;
+  }
+  /**
+   *  Returns the minimum value recorded. If size of list is zero, returns NaN.
+   *
+   *  @return The minimum value recorded, or NaN if size of data is zero.
+  **/
+  double min()
+  {
+    return getStatistics().min;
+  }
+  /**
+   *  Returns the standard deviation (population) of all data recorded. If size of list is zero, returns NaN.
```

Comment:
80 cols?

Reply:
Hm, none of the linters have complained. I'll try rerunning and auto formatting.